packages/compass/src/main/auto-update-manager.ts

```diff
@@ -155,67 +168,125 @@ const checkForUpdates: StateUpdateAction = async function checkForUpdates(
   }
 };
 
-const disableAutoUpdates: StateUpdateAction = function disableAutoUpdates() {
+const disableAutoUpdates: StateEnterAction = function disableAutoUpdates() {
   log.info(
     mongoLogId(1_001_000_162),
     'AutoUpdateManager',
     'Disabling auto updates'
   );
 };
 
+interface AutoUpdateManagerStateDefinition {
+  nextStates: AutoUpdateManagerState[];
+  enter: StateEnterAction;
+}
+
+const noop = () => {
+  /* noop */
+};
+
 /**
- * State update map defined as `{ fromState: toState() }` where toState is the
- * action that needs to happen on state transition
+ * State update map defined as `{ state: [fromStates, enter()]}` where `enter` is
+ * the action that needs to happen when entering the respective state.
  */
-const STATE_UPDATE: Partial<
-  Record<
-    AutoUpdateManagerState,
-    Partial<
-      Record<
-        AutoUpdateManagerState,
-        (
-          this: { maybeInterrupt(): void | never },
-          updateManager: typeof CompassAutoUpdateManager,
-          ...extraArgs: any[]
-        ) => void | Promise<void>
-      >
-    >
-  >
+const STATE_UPDATE: Record<
+  AutoUpdateManagerState,
+  AutoUpdateManagerStateDefinition
```

Comment:
(heads up that Github hides the diff for this file by default)

This flips around the logic a bit; instead of defining specific actions to be taken when going from A to B (and the allowed set of state transitions being implied by the presence of an explicitly defined action), this defines an action to be performed when *entering* state B, and a valid set of states to transition to.

I hope this is okay (@gribnoysup you wrote this, so I'd like to check with you), but I'll have to say, I kind of had to do this in order for me to really feel like I covered the right set of changes here 😅 I think the fact that we had previously factored out functions like `manualCheck`, `checkForUpdates`, `disableAutoUpdates` so that we were able to use them for multiple [from, to] pairs was a sign to me that this would make sense, obviously we could keep following that pattern if we really felt like we want to keep it.

Reply:
Yeah, no strong preference here on how the state transitions are done. I did like the explicitness of the old state transition set up, but I also agree that it was hard to follow and make sure that there are no missing transitions. I like the change 🙂